core/src/main/scala/tech/beshu/ror/accesscontrol/blocks/rules/LdapAuthorizationRule.scala

```diff
@@ -75,6 +75,34 @@ object LdapAuthorizationRule {
   }
 
   final case class Settings(ldap: LdapAuthorizationService,
-                            permittedGroups: UniqueNonEmptyList[Group],
+                            permittedGroups: GroupsLogic,
                             allLdapGroups: UniqueNonEmptyList[Group])
+
+  sealed trait GroupsLogic {
+    def groups: UniqueNonEmptyList[Group]
+
+    def availableGroupsFrom(userGroups: Set[Group], ruleGroups: Set[Group]): Option[UniqueNonEmptyList[Group]]
```

Comment:
IMO it's better not to add methods to ADT that won't be overridden as val (see the example above). IMO it's better to create a typeclass (or extension method class). Eg.

```
implicit class GroupsLogicExecutor(val groupsLogic: GroupsLogic) extends AnyVal {
  def availableGroupsFrom(userGroups: Set[Group], ruleGroups: Set[Group]): Option[UniqueNonEmptyList[Group]] = {
    groupsLogic match {
      case And(_) => [do sth]
      case Or(_) => [do sth different]
    }
  }
}
```

then you will be able to use it in the same way:
```
val logic: GroupsLogic = ???
logic.availableGroupsFrom(...)
```

Btw. does the `availableGroupsFrom` needs two parameters? You have permitted groups in the GroupsLogic instance